c/misra/src/rules/RULE-17-11/FunctionWithNoReturningBranchShouldBeNoreturn.ql

```diff
@@ -0,0 +1,24 @@
+/**
+ * @id c/misra/function-with-no-returning-branch-should-be-noreturn
+ * @name RULE-17-11: A function without a branch that returns shall be declared with _Noreturn
+ * @description Functions which cannot return should be declared with _Noreturn.
+ * @kind problem
+ * @precision high
+ * @problem.severity recommendation
+ * @tags external/misra/id/rule-17-11
+ *       external/misra/obligation/advisory
+ */
+
+import cpp
+import codingstandards.c.misra
+import codingstandards.c.Noreturn
+
+from Function f
+where
+  not isExcluded(f, NoReturnPackage::returnStatementInNoreturnFunctionQuery()) and
+  not f instanceof NoreturnFunction and
+  not mayReturn(f) and
+  f.hasDefinition() and
+  f.getName() != "main" // Allowed exception; _Noreturn main() is undefined behavior.
```

Comment:
I would generally recommend the use of `not X = Y` instead of `X != Y`, because I tend to find the semantics are a better match for my intentions when I write this type of formula.

As a concrete example, imagine `File` had a predicate called `getALanguage()`, and we wanted to filter on files which were not one of `cpp` or `c`. With `!=` we might write:
```codeql
from File f
where f.getALanguage() != ["cpp", "c"]
select f
```
Somewhat surprisingly, if a file was compiled as both `c` and `cpp` (as can happen with, say, a header file), then this will still report it! This is because `!=` essentially says there is at least one pair of left and right for which the two are not equal. If we instead used `where not f.getALanguage() = ["cpp", "c"]` then we would get our expected behaviour - this is because `not .. = ..` essentially says that there doesn't exist a pair of left and right for which the two are equal.

In the specific case of this query, I don't think it matters - each function has one name, and is being compared to one named value - but getting into the habit of using `!=` means you may end up using it in a situation where it's not appropriate in the future.

More on this in the language reference:
https://codeql.github.com/docs/ql-language-reference/formulas/#footnote-1

```suggestion
  not f.getName() = "main" // Allowed exception; _Noreturn main() is undefined behavior.
```

Reply:
Thanks for the detailed explanation! I thought I had handled this correctly...but I was only worrying about the case of `not exists(getName())`, and _not_ the case where a function has multiple names! Thanks! :brain: